internal/server/v1/observations/derived_series.go

```diff
@@ -0,0 +1,507 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package observations implements for /v1/derived/observations/series/...
+package observations
+
+import (
+	"context"
+	"fmt"
+	"go/ast"
+	"go/parser"
+	"go/token"
+	"os"
+	"reflect"
+	"sort"
+	"strings"
+
+	pb "github.com/datacommonsorg/mixer/internal/proto"
+	"github.com/datacommonsorg/mixer/internal/server/stat"
+	"github.com/datacommonsorg/mixer/internal/store"
+)
+
+// DerivedSeries implements API for Mixer.DerivedObservationsSeries.
+func DerivedSeries(
+	ctx context.Context,
+	in *pb.DerivedObservationsSeriesRequest,
+	store *store.Store,
+) (*pb.DerivedObservationsSeriesResponse, error) {
+	resp := &pb.DerivedObservationsSeriesResponse{}
+	entity := in.GetEntity()
+
+	// Parse the formula to extract all the variables, used for reading data from BT.
+	varCalculator, err := newVarCalculator(in.GetFormula())
+	if err != nil {
+		return resp, err
+	}
+	statVars := varCalculator.statVars()
+
+	// Read data from BT.
+	btData, err := stat.ReadStatsPb(
+		ctx, store.BtGroup, []string{entity}, statVars)
+	if err != nil {
+		return resp, err
+	}
+	entityData, ok := btData[entity]
+	if !ok {
+		return resp, err
+	}
+
+	// Calculate.
+	result, err := varCalculator.calculate(entityData)
+	if err != nil {
+		return resp, err
+	}
+	for _, p := range result.points {
+		resp.Observations = append(resp.Observations, &pb.PointStat{
+			Date:  p.GetDate(),
+			Value: p.GetValue(),
+		})
+	}
+
+	return resp, nil
+}
+
+type varCalculatorSeries struct {
+	statMetadata *pb.StatMetadata
+	// Sorted by date.
+	points []*pb.PointStat
+}
+
+// The date key of a time series is concatenation of all sorted dates.
+func (s *varCalculatorSeries) dateKey() string {
+	dates := []string{}
+	for _, point := range s.points {
+		dates = append(dates, point.GetDate())
+	}
+	return strings.Join(dates, "")
+}
+
+type varInfo struct {
+	statVar          string
+	statMetadata     *pb.StatMetadata
+	seriesCandidates []*varCalculatorSeries
```

Comment:
each series in seriesCandidates already has statMetadata?

Reply:
In the future, when we swtich from varCalculatorSeries to pb.SourceSeries, yes it'll have statMetadata. For now I'll keep using varCalculatorSeries since pb.SourceSeries doesn't have ordered 'val'.